controller/state.go

```diff
@@ -305,6 +308,50 @@ func verifyGnuPGSignature(revision string, project *appv1.AppProject, manifestIn
 	return conditions
 }
 
+func (m *appStateManager) diffArrayCached(configArray, liveArray []*unstructured.Unstructured, opts ...diff.Option) (*diff.DiffResultList, error) {
+	numItems := len(configArray)
+	if len(liveArray) != numItems {
+		return nil, fmt.Errorf("left and right arrays have mismatched lengths")
+	}
+
+	managedResources := make([]*appv1.ResourceDiff, 0)
+	if m.cache.GetAppManagedResources("", &managedResources) != nil {
+		// (rare) cache miss
+		return diff.DiffArray(configArray, liveArray, opts...)
+	}
+
+	diffResultList := diff.DiffResultList{
+		Diffs: make([]diff.DiffResult, numItems),
+	}
+
+	for i := 0; i < numItems; i++ {
+		var dr *diff.DiffResult
+		if i < len(managedResources) {
```

Comment:
It is not safe assume that the matching resource is stored in the exact same position in cache. Instead we need match resoures by namespace/name/group/kind . I would suggest to group resources by storing them in `map[kube.ResourceKey]*unstructured.Unstructured`